multicluster/controllers/multicluster/gateway_controller.go

```diff
@@ -113,6 +113,11 @@ func (r *GatewayReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
 			}
 			return nil
 		}
+
+		if !existingResExport.DeletionTimestamp.IsZero() {
+			return fmt.Errorf("existing ResourceExport is being deleted, retry later")
```

Comment:
Question - if we do not add this check, will updateResourceExport() fail? Then what is the difference of returning an error here, compared to letting updateResourceExport() fail and return an error at line 122?

Have you checked how long the new active GW can be finally recreated, when the race condition happens and we have to retry here?

Reply:
updateResourceExport() will not fail because update is allowed when DeletionTimestamp is not empty. The updated one will be recycled by K8s because DeletionTimestamp is not zero after that.
The new active Gateway is created immediately when old Gateway is removed or down. The corresponding ClusterInfo type of ResourceExport will also be created almost immediately because it's usually one time retry unless the leader controller is down.

Reply:
In this case, could there be another race condition that the check is passed here, but DeletionTimeStamp is set before or after update is done?

Reply:
No, we only have one goroutine in gateway controller to handle Gateway change considering Gateway event rarely happens.

Reply:
But you get the ResourceExport from client cache right? The state can be stale?

 When the Gateway changes, should we create the ResourceExport rather than update it?

Reply:
Yes, we actually will recreate the ClusterInfo type of ResourceExport when the Gateway Node changes. Update only happens when the Gateway Node is not changed but Gateway IP is changed.

Reply:
It is still not clear what we are fixing with this commit. You said in the commit description:

> But there is a case that a new ClusterInfo of ResourceExport creation may fail when the leader controller process is slow and existing ResourceExport is not deleted in time. Fix the issue through retry when the existing ResourceExport's DeletionTimestamp is not zero.

If creation will fail, then the controller anyway will retry right? Why we need to add a separate check and retry?

Reply:
Hi @jianjuns , I added a few new descriptions in the summary, could you take a look? thanks

Reply:
Ok, we then get back to my earlier comment. Could you answer the questions?

> But you get the ResourceExport from client cache right? The state can be stale?

> When the Gateway changes, should we create the ResourceExport rather than update it?

Reply:
I recalled that we discussed this before, and the solution is not comparing existing Gateway with the Gateway info in the ResourceExport. The controller will always update Gateway IP ignoring the value in the existing ResourceExport, and the Gateway change will be handled by sequence, so I suppose it's Ok not to recreate it?

Reply:
Did not remember what we discussed before, but at that time we did not discover the ResourceExport issue right? I am asking your current code gets the ResourceExport and checks ResourceExport, but what will happen if the ResourceExport you read (could be from client cache) is stale?

Could you just answer that question? I feel our discussion is rather inefficient. We went a long loop, but never came back to the original point:
> But you get the ResourceExport from client cache right? The state can be stale?

Maybe let us talk offline.

Reply:
Yes, there is a chance that cache might be stale, but it happens rarely. I didn't see this kind of issue yet. And the reconcile logic on our controller is responsible to make sure that information will be eventually correct. Here is a few description on the [controller-runtime FAQ](https://github.com/kubernetes-sigs/controller-runtime/blob/master/FAQ.md#q-my-cache-might-be-stale-if-i-read-from-a-cache-how-should-i-deal-with-that). 

And this bug is not introduced by stale ResourceExport from cache. The cache is correct. cahce should contains the deleting ResourceExport when the DeletionTimestamp is not zero.  It's expected to get deleting objects before K8s do the recycle process. `kubectl get` will print them as well. 

Before Gateway HA, we didn't delete ClusterInfo type of ResourceExport unless there is no Gateway at all, it always updates the corresponding ResourceExport as long as there is an active Gateway even there is a Gateway Node change. After Gateway HA, this issue shows up only when leader controller is slow(the Finalizer on ResourceExport will be removed only after corresponding ResourceImport is deleted, then K8s recycle happens.), or K8s recycle is not finished in time. I missed this kind of issue when I verify Gateway HA manually, but found this in a new environment during networkPolicyOnly mode testing.

Let me know if you'd like to discuss more details offline. thanks.

Reply:
Could you elaborate how the reconciliation can eventually fix the issue, if the ResourceExport is deleted due to stale client cache?

We should never assume a corner case is rare and so do not handle it. What can happen will happen in CS.

Reply:
Hi, @jianjuns , the client cache may have three different cases after ClusterInfo of ResourceExport in the leader cluster is deleted.
1. Old ResourceExport is still existing on the client cache and the DeletionTimestamp is zero
2. Old ResourceExport is still existing on the client cache and the DeletionTimestamp is not zero
3. Old ResourceExport has been removed from current cache
The 3rd will go through correct process without any issue. The 2nd one can be fixed by current PR to retry. But for the 1st case, when a new Gateway coming, it will go to update existing ResourceExport without any issue. I didn't find there is a valid way to handle this case unless we disable cache for ResourceExport and get objects from leader API server directly.
The 1st case is rare but there is indeed a chance to happen considering there are actually two clusters and cache maybe built up via internet. The controller-runtime allow user to avoid cache for specific objects, so I suppose we can completely avoid this kind of issue by disabling cache for ResourceExport.

Reply:
Hi Lan, I do not know if it is a good practice to disable client cache, and if we should use that as a common approach for such race conditions. I would like to get @tnqn's comment.

I do feel there are two other ways for this particular problem:
1. When DeletionTimestamp != 0, do not return and retry, but treat ResourceExport as already deleted, and change update to create.
2. Maintain a state in controller about the ClusterInfo Export (or say Gateway state of ClusterInfo Export), and based on to decide what to do, rather than getting & checking ResourceExport.

Reply:
Sure, I will check and refine this PR with one of your suggested solutions for this particular issue first. thanks.

Reply:
Disabling client cache won't resolve it as the object in apiserver may be updated after the API is called. 
This situation is very common and a typical way to resolve it is, using the object in the cache without concerning if it's stale or not, updating the object with resourceVersion got from the cache (clientset automatically does it as long as the object has resourceVersion), then if the object is stale, apiserver will reject the update request with updateConflict error, and client should retrieve the latest version from apiserver directly and retry. Example: https://github.com/antrea-io/antrea/blob/main/pkg/agent/controller/egress/egress_controller.go#L561

But I wonder even if it updates the resource export successfully, could the other slow leader delete it because it thinks this is the one for itself? If yes, the deletion logic should check whether the object matches and set resourceVersion when sending the delete request.

Reply:
@jianjuns  I have refined the code to create ResourceExport when DeletionTimestamp != 0. After checking existing codes, it already has resourceVersion when it's trying to update existing ResourceExport, so it should fail when there is conflict and retry. For the deletion action, I feel it's Ok without resourceVersion considering Gateway webhook will guarantee that only one Gateway is allowed in a member cluster. all events should happen by sequence.

Reply:
Do we still keep the resourceVersion in update or not? If yes, maybe add a comment before the update call about that.

@tnqn : the controller has only a single worker, so no synchronization issue or reordering.

Reply:
@jianjuns yes, the `updateResourceExport` will reuse existing ResourceExport's resourceVersion to update. I have added a comment for this.